agent/src/softwarecontaineragent.cpp

```diff
@@ -0,0 +1,231 @@
+#include "softwarecontaineragent.h"
+
+SoftwareContainerAgent::SoftwareContainerAgent(Glib::RefPtr<Glib::MainContext> mainLoopContext, int preloadCount, bool shutdownContainers, int shutdownTimeout)
+    : m_mainLoopContext(mainLoopContext)
+    , m_preloadCount(preloadCount)
+    , m_shutdownContainers(shutdownContainers)
+{
+    triggerPreload();
+    m_softwarecontainerWorkspace.m_containerShutdownTimeout = shutdownTimeout;
+}
+
+SoftwareContainerAgent::~SoftwareContainerAgent()
+{
+}
+
+void SoftwareContainerAgent::triggerPreload()
+{
+    //        log_debug() << "triggerPreload " << m_preloadCount - m_preloadedContainers.size();
+    while (m_preloadedContainers.size() != m_preloadCount) {
+        auto container = new SoftwareContainerLib(m_softwarecontainerWorkspace);
+        container->setContainerIDPrefix("Preload-");
+        container->preload();
+        m_preloadedContainers.push_back(SoftwareContainerLibPtr(container));
+    }
+}
+
+void SoftwareContainerAgent::deleteContainer(ContainerID containerID)
+{
+    bool valid = ((containerID < m_containers.size()) && (m_containers[containerID] != nullptr));
+    if (valid) {
+        m_containers[containerID] = nullptr;
+    } else {
+        log_error() << "Invalid container ID " << containerID;
+    }
+}
+
+bool SoftwareContainerAgent::checkContainer(ContainerID containerID, SoftwareContainerLib *&container)
+{
+    bool valid = ((containerID < m_containers.size()) && (m_containers[containerID] != nullptr));
+    if (valid) {
+        container = m_containers[containerID].get();
+    } else {
+        log_error() << "Invalid container ID " << containerID;
+    }
+
+    return valid;
+}
+
+ReturnCode SoftwareContainerAgent::readConfigElement(const json_t *element)
+{
+    std::string wo;
+    if(!read(element, "writeOften", wo))
+    {
+        log_debug() << "writeOften not found";
+        m_softwarecontainerWorkspace.m_writeOften = false;
+    } else {
+        if (wo == "1")
+        {
+            m_softwarecontainerWorkspace.m_writeOften = true;
+        } else {
+            m_softwarecontainerWorkspace.m_writeOften = false;
+        }
+    }
+    return ReturnCode::SUCCESS;
+}
+
+bool SoftwareContainerAgent::parseConfig(const std::__cxx11::string &config)
+{
+    if (config.size() == 0)
+    {
```

Comment:
If brackets should not be on the line after then also fix them from line 132 and forward.